padding/padding.go

```diff
@@ -0,0 +1,157 @@
+	return msg - leak
+}
+
+//Inputs:
+//msgLen int, Is the length of the message to be padded.
+//overhead int, is how much overhead the encryption,
+//and padding will add to the length of the message
+
+//Returns how many bytes of padding need to be added
+//Works by masking the non zero bits to 1, then inverting and adding 1.
+//Another way to do this would just be look at each bit 1 at a time
+//For zeroBits
+// sum+= l & (1<<i)
+//Code so that if l&(1<<i) is 0 it adds 2^i, otherwise nothing
+func GetPaddingLen(msgLen, overhead uint64) uint64 {
+	paddingNeeded := uint64(msgLen + overhead)
+	zeroBits := GetZeroBits(msgLen, overhead)
+
+	var i, mask uint64
+	//Generate a mask that we use to isolate the zeroBits bits of the length
+	for i = 0; i < zeroBits; i++ {
+		mask |= (1 << i)
+		//fmt.Println(i, zeroBits)
+	}
+	//invert the mask
+	mask = ^mask
+	//Or the mask with our length. Now all non zero bits are 1
+	paddingNeeded |= mask
+	//invert l, now we have the ^l, ^l + l will give us only ones
+	//So if we add 1 to that we get 0 in all of the zero bits.
+	//And the next largest bit will be a 1.
+	paddingNeeded = ^paddingNeeded
+	paddingNeeded = paddingNeeded + 1
+	//This is the case where all of the zero bits are already 0.
+	if float64(paddingNeeded) == math.Pow(2, float64(zeroBits)) {
+		paddingNeeded = 0
+	}
+	return paddingNeeded
+}
+
+//Function that checks if a message is padded correctly
+//Inputs:
+//msgLen int, msgLen is the length of the message to check.
+//Returns true if it is an acceptable length, false otherwise.
+func CheckZeroBits(msgLen uint64) bool {
+	paddingNeeded := uint64(msgLen)
+	zeroBits := GetZeroBits(msgLen, 0)
+	var i, mask uint64
+	for i = 0; i < zeroBits; i++ {
+		mask |= (1 << i)
+	}
+	mask = ^mask
+	paddingNeeded |= mask
+	paddingNeeded = ^paddingNeeded
+	paddingNeeded = paddingNeeded + 1
+	//This is the case were the message is padded correctly
+	//all bits are 0 -> all are 1 -> +1 = zeroBits^2
+	if float64(paddingNeeded) == math.Pow(2, float64(zeroBits)) {
+		paddingNeeded = 0
+	}
+	if paddingNeeded == 0 {
+		return true
+	}
+	return false
+}
+
+//generates padding, can be later modified if something else is better.
+//There is also probably a more efficient way to generate it.
+
+//Inputs:
+//paddingAmount uint64 --this is the number of bytes of padding to generate.
+
+func GeneratePadding(paddingAmount uint64) []byte {
+	padding := make([]byte, paddingAmount)
+	var i uint64
+	for i = 0; i < paddingAmount; i++ {
+		padding[i] = PADBYTE //random choice
+	}
+	return padding
+}
+
+//Inputs:
+//overhead int, is the number of bytes the encryption will add to the plaintext
+//msg []byte, is the plaintext message that needs to be padded.
+//Returns a properly padded message.
+func PadGeneric(msg []byte, overhead uint64) []byte {
+	padAmount := GetPaddingLen(uint64(len(msg)), overhead+PADLEN)
+	padding := GeneratePadding(padAmount)
+	buf := new(bytes.Buffer)
+	binary.Write(buf, binary.BigEndian, int64(padAmount))
```

Comment:
Almost certainly not, but I am not sure what value would make the most sense. possibly just shrink it to int32? Although if this is used for a file bigger then 8 GB it is possible it would need more padding. Although it could be an application dependent thing probably.